query-grammar/src/query_grammar.rs

```diff
@@ -374,6 +374,25 @@ pub fn parse_to_ast<'a>() -> impl Parser<&'a str, Output = UserInputAst> {
     spaces()
         .with(optional(ast()).skip(eof()))
         .map(|opt_ast| opt_ast.unwrap_or_else(UserInputAst::empty_query))
+        .map(|ast| rewrite_ast(ast))
+}
+
+/// Removes unnecessary children clauses in AST
+///
+/// Motivated by [issue #1433](https://github.com/quickwit-oss/tantivy/issues/1433)
+fn rewrite_ast(input: UserInputAst) -> UserInputAst {
+    if let UserInputAst::Clause(terms) = input {
+        UserInputAst::Clause(terms.into_iter().map(rewrite_ast_clause).collect())
```

Comment:
I understand that `Vec::into_iter().collect::<Vec<_>>()` can be optimized to act in-place instead of re-allocating. Can we rely on this here and do we want to? Which downsides do you see for something like

```rust
if let UserInputAst::Clause(terms) = &mut input {
  for term in terms {
    rewrite_ast_clause(&mut term);
  }
}
```

Reply:
> Can we rely on this here and do we want to?

It's hard yo say. If this code is on a hot path, it's better to be safe I guess.

> Which downsides do you see for something like

Apart from it's maybe kinda not so idiomatic in Rust I can't think of any.

Rewrote `rewrite_ast_clause()` as `fn(&mut (Option<Occur>, UserInputAst))`